pkg/gadgets/socket-collector/tracer/tracer.go

```diff
@@ -0,0 +1,268 @@
+	case TCP_LISTEN:
+		return "LISTEN"
+	case TCP_CLOSING:
+		return "CLOSING"
+	case TCP_NEW_SYN_RECV:
+		return "NEW_SYN_RECV"
+	default:
+		prifixedLog(fmt.Sprintf("Invalid TCP status: %s (%v)", statusHex, statusBytes))
+	}
+
+	return "UNKNOWN"
+}
+
+func parseUDPStatus(statusHex string) string {
+	statusBytes, err := hex.DecodeString(statusHex)
+	if err != nil {
+		prifixedLog(fmt.Sprintf("Failed to decode hex-string UDP status %q: %s", statusHex, err))
+		return ""
+	}
+
+	switch uint8(statusBytes[0]) {
+	case TCP_ESTABLISHED:
+		return "ACTIVE"
+	case TCP_CLOSE:
+		return "INACTIVE"
+	default:
+		prifixedLog(fmt.Sprintf("Unexpected UDP status: %s (%v)", statusHex, statusBytes))
+	}
+
+	return "UNKNOWN"
+}
+
+func getTCPIter() *link.Iter {
+	objs := IterTCPv4Objects{}
+	if err := LoadIterTCPv4Objects(&objs, nil); err != nil {
+		prifixedLog(fmt.Sprintf("Failed to load TCP BPF objects: %v", err))
+		return nil
+	}
+	defer objs.Close()
+
+	it, err := link.AttachIter(link.IterOptions{
+		Program: objs.DumpTcp4,
+	})
+	if err != nil {
+		prifixedLog(fmt.Sprintf("Failed to attach TCP BPF iterator: %v", err))
+		return nil
+	}
+
+	return it
+}
+
+func getUDPIter() *link.Iter {
+	objs := IterUDPv4Objects{}
+	if err := LoadIterUDPv4Objects(&objs, nil); err != nil {
+		prifixedLog(fmt.Sprintf("Failed to load UDP BPF objects: %v", err))
+		return nil
+	}
+	defer objs.Close()
+
+	it, err := link.AttachIter(link.IterOptions{
+		Program: objs.DumpUdp4,
+	})
+	if err != nil {
+		prifixedLog(fmt.Sprintf("Failed to attach UDP BPF iterator: %v", err))
+		return nil
+	}
+
+	return it
+}
+
+func RunCollector(pid uint32, filter *gadgetv1alpha1.ContainerFilter) (string, error) {
+	if err := increaseRlimit(); err != nil {
+		return "", fmt.Errorf("failed to increase memlock limit: %s", err)
+	}
+
+	// TODO: Use filters to collect only TCP or only UDP information
+	iters := []*link.Iter{}
+	iters = append(iters, getTCPIter())
+	iters = append(iters, getUDPIter())
+
+	// TCP/UDP BPF Iters read information from the current namespace thus
+	// we need to change network namespace before opering the iterator
+	netnsHandle, err := netns.GetFromPid(int(pid))
+	if err != nil {
+		return "", fmt.Errorf("failed to retrieve network namespace from PID %d: %w", pid, err)
+	}
+	defer netnsHandle.Close()
+
+	if err := netns.Set(netnsHandle); err != nil {
+		return "", fmt.Errorf("failed to set network namespace of PID %d (NetNsHandle %q): %w", pid, netnsHandle.String(), err)
+	}
+
+	var data []Socket
+
+	for _, it := range iters {
+		if it == nil {
+			continue
+		}
+
+		fd, err := it.Open()
+		if err != nil {
+			prifixedLog(fmt.Sprintf("Failed to open BPF iterator: %v", err))
+			continue
+		}
+		defer fd.Close()
+
+		scanner := bufio.NewScanner(fd)
+		for scanner.Scan() {
+			var status string
+
+			fields := strings.Fields(scanner.Text())
+			if len(fields) != 6 {
+				prifixedLog(fmt.Sprintf("Invalid fields length to describe socket: %v", fields))
+				continue
+			}
```

Comment:
I think we can simply the logic to convert it back by using [Sscanf](https://pkg.go.dev/fmt#Sscanf). You know the format you used to convert it to string (`"tcp %08X %04X %08X %04X %02X\n"`) can do revert that by using sscanf.

This is a sort of pseudocode (C + golang mix) that would do it:

```
uint16 destp, srcp;
uint32 dest, src;
string type_;
uint8 status; 
if fmt.Sscanf(scanner.Text(), &type_, &src, &srcp, &dest, &destp, &status) != 6 {
   return err...
}
```

After it you have some of the types already available (ports), and it's quite straightforward to convert others as ips (you then avoid to use hex.DecodeString too).

Reply:
You're right, I will use this approach. I don't know why I complicated myself like this :sweat_smile:.

Reply:
After moving to sscanf, the only parameters that still need to be parsed are the IPs. I am wondering if there is another way I don't neither have to specify the endianness nor do it manually:
```
func parseIPv4(ipU32 uint32) string {
	ipBytes := make([]byte, 4)

	binary.BigEndian.PutUint32(ipBytes, ipU32)
	ip := net.IP(ipBytes)

	return ip.String()
}
```
The IPv6 should be identical but using uint64 and a 16-byte slice.

Reply:
I'm not aware of other ways to do it.